app/metal-metadata-server/main.go

```diff
@@ -147,406 +153,431 @@ func FetchConfig(w http.ResponseWriter, r *http.Request) {
 			continue
 		}
 
-		// If ref matches, fetch the bootstrap data from machine resource that owns this metal machine
 		if serverRefString == uuid {
-			ownerList, present, err := unstructured.NestedSlice(metalMachine.Object, "metadata", "ownerReferences")
-			if err != nil {
-				throwError(
-					w,
-					http.StatusInternalServerError,
-					fmt.Errorf(
-						"failure fetching ownerRefs from metal machine %s/%s: %s",
-						metalMachine.GetNamespace(),
-						metalMachine.GetName(),
-						err,
-					),
-				)
-
-				return
-			}
+			numMatchingMachines++
 
-			if !present {
-				throwError(
-					w,
-					http.StatusNotFound,
-					fmt.Errorf(
-						"ownerRefList not found for metalMachine",
-					),
-				)
+			matchingMetalMachine = metalMachine
+		}
+	}
 
-				return
-			}
+	// bail early if we have multiple matches or no matches
+	switch {
+	case numMatchingMachines > 1:
+		throwError(
+			w,
+			http.StatusNotFound,
+			fmt.Errorf(
+				"multiple matching metal machines for uuid %q, possible orphaned cluster",
+				uuid,
+			),
+		)
 
-			var ownerRef *metav1.OwnerReference
+		return
+	case numMatchingMachines == 0:
+		throwError(
+			w,
+			http.StatusNotFound,
+			fmt.Errorf(
+				"failure finding matching metal machine for uuid: %q",
+				uuid,
+			),
+		)
 
-			for _, ownerItem := range ownerList {
-				tempOwnerRef := &metav1.OwnerReference{}
+		return
+	}
 
-				err = runtime.DefaultUnstructuredConverter.FromUnstructured(ownerItem.(map[string]interface{}), tempOwnerRef)
-				if err != nil {
-					throwError(
-						w,
-						http.StatusInternalServerError,
-						fmt.Errorf("failure converting ownerItem to ownerRef: %s", err),
-					)
+	// If we have a match, fetch the bootstrap data from machine resource that owns this metal machine
+	ownerList, present, err := unstructured.NestedSlice(matchingMetalMachine.Object, "metadata", "ownerReferences")
```

Comment:
From here down looks so much worse that it is. It's really just an un-indentation from moving the for loop that goes through metal machines.